star_ratings/templatetags/ratings.py

```diff
@@ -39,6 +40,7 @@ def ratings(context, item, icon_height=app_settings.STAR_RATINGS_STAR_HEIGHT, ic
     # template name can be passed as a template parameter
     template_name = template_name or context.get('star_ratings_template_name') or 'star_ratings/widget.html'
     return loader.get_template(template_name).render({
+        'csrf_token': csrf(request)['csrf_token'],
```

Comment:
Normally we should let the renderer generate its own `RequestContext` object as the error suggests, the reason it isn't is that we aren't passing the request to the render function. I think rather than setting the csrf token directly we should instead be passing the request to the `Template.render` function. This will make sure we don't miss any other properties we get from the request context. Something like:

```
loader.get_template(template_name).render(
    {...}, 
    request=request
)
```

We can also remove `request` from our custom context too.

Reply:
Yes, you are right, it makes sense to pass the request directly to the `Template.render` function and remove `'request': request` from the custom context. But, if I change the code to reflect this, one test in `./scripts/runtests.py` fails:

> tests/test_template_tags_ratings.py:68: in test_request_is_added_to_the_result
>     self.assertEqual(request, context['request'])

The test fails, because I removed `'request': request` from the custom context. This raises following questions in my mind. Should I alter/remove the test? Does maybe others people code depend on request being found in the custom context? Or shall I pass the request directly to the `Template.render` function, but leave `'request': request` in the custom context as well?

What do you think?

Reply:
The request should be added by the context processor `django.template.context_processors.request` so I'm a little confused whats going on here.

As for the tests, we shouldn't alter the tests as this is behaviour the app and probably others depend on. For now if we leave the `'request': request` in the custom context and we can close this down so the csrf issue is fixed and we can look into the context behaviour later (feel free to take a look if you like).

Thanks for this :smile:

Reply:
I changed the code to leave `'request': request` and pass the request to the `Template.render ` function.

I also had a deeper look at the warning, which only popped up, when the user was logged in:

`[...]/venv/lib/python3.6/site-packages/django/template/defaulttags.py:63: UserWarning: A {% csrf_token %} was used in a template, but the context did not provide the value.  This is usually caused by not using RequestContext.`

It looks like the contents (e.g. csrf token) of `'request': request` in the custom context is never processed via `django.template.context_processors.request`, because if `request=request` is missing in the `Template.render` function, `Context` instead of `RenderContext` is used in the `make_context` function (called by the `Template.render` function).

Without adding `request=request` to the `Template.render` function, there is no csrf token added to the template (even if there is a `'request': request` in the custom context). But the check, if a csrf token is included in the template tag is only done by `CsrfTokenNode` in`template.defaulttags`, if a user is authenticated und therefore the warning did only appear if a user was logged in.